src/monodroid/jni/embedded-assemblies.cc

```diff
@@ -299,50 +434,260 @@ EmbeddedAssemblies::gather_bundled_assemblies_from_apk (const char* apk, monodro
+		return false;
+	}
+
+	return true;
+}
+
+bool
+EmbeddedAssemblies::typemap_load_file (int dir_fd, const char *dir_path, const char *file_path, TypeMapModule &module)
+{
+	log_debug (LOG_ASSEMBLY, "typemap: loading TypeMap file '%s/%s'", dir_path, file_path);
+
+	bool ret = true;
+	BinaryTypeMapHeader header;
+	size_t file_size;
+	int fd = -1;
+
+	module.java_map = nullptr;
+	module.map = nullptr;
+	module.duplicate_map = nullptr;
+
+	if (!typemap_read_header (dir_fd, "TypeMap", dir_path, file_path, MODULE_MAGIC, header, file_size, fd)) {
+		ret = false;
+		goto cleanup;
+	}
+
+	ret = typemap_load_file (header, dir_path, file_path, fd, module);
+
+  cleanup:
+	if (fd >= 0)
+		close (fd);
+
+	if (!ret) {
+		delete[] module.java_map;
+		module.java_map = nullptr;
+		delete[] module.map;
+		module.map = nullptr;
+		delete[] module.duplicate_map;
+		module.duplicate_map = nullptr;
+	}
+
+	return ret;
+}
+
 void
 EmbeddedAssemblies::try_load_typemaps_from_directory (const char *path)
 {
-	// read the entire typemap file into a string
-	// process the string using the add_type_mapping
-	char *dir_path = utils.path_combine (path, "typemaps");
-	if (dir_path == nullptr || !utils.directory_exists (dir_path)) {
-		log_warn (LOG_DEFAULT, "directory does not exist: `%s`", dir_path);
-		free (dir_path);
+	if (!application_config.instant_run_enabled) {
+		log_info (LOG_ASSEMBLY, "typemap: instant run disabled, not loading type maps from storage");
 		return;
 	}
 
+	simple_pointer_guard<char[]> dir_path (utils.path_combine (path, "typemaps"));
 	monodroid_dir_t *dir;
 	if ((dir = utils.monodroid_opendir (dir_path)) == nullptr) {
-		log_warn (LOG_DEFAULT, "could not open directory: `%s`", dir_path);
-		free (dir_path);
+		log_warn (LOG_DEFAULT, "typemap: could not open directory: `%s`", dir_path.get ());
 		return;
 	}
 
-	monodroid_dirent_t *e;
-	while ((e = androidSystem.readdir (dir)) != nullptr) {
-#if WINDOWS
-		char *file_name = utils.utf16_to_utf8 (e->d_name);
-#else   /* def WINDOWS */
-		char *file_name = e->d_name;
-#endif  /* ndef WINDOWS */
-		char *file_path = utils.path_combine (dir_path, file_name);
-		if (utils.monodroid_dirent_hasextension (e, ".mj") || utils.monodroid_dirent_hasextension (e, ".jm")) {
-			char *val = nullptr;
-			size_t len = androidSystem.monodroid_read_file_into_memory (file_path, val);
-			if (len > 0 && val != nullptr) {
-				if (utils.monodroid_dirent_hasextension (e, ".mj")) {
-					if (!add_type_mapping (&managed_to_java_maps, file_path, override_typemap_entry_name, ((const char*)val)))
-						delete[] val;
-				} else if (utils.monodroid_dirent_hasextension (e, ".jm")) {
-					if (!add_type_mapping (&java_to_managed_maps, file_path, override_typemap_entry_name, ((const char*)val)))
-						delete[] val;
-				}
-			}
+	int dir_fd;
+#if __ANDROID_API__ < 21
+	dir_fd = -1;
+#else
+	dir_fd = dirfd (dir);
+#endif
+
+	constexpr char index_name[] = "typemap.index";
```

Comment:
We only support a *single* `typemap.index` file when "instant run" is enabled?

Could this instead support *multiple* files within the `typemaps` directory?  That way we could (eventually) have a typemap file per-assembly, instead of one file for the entire app, which should help with incremental builds.

Reply:
We need just a single **index** file. We support many `*.typemap`, this PR implements typemap file per-assembly (or rather, per assembly module as it is possible some assemblies may have multiple modules)

Reply:
I'm still not sure I grok an "index" file, and why we need a single *anything*.  Perhaps this is for mapping assembly names to MVIDs?  (If so, `.mvidmap` might make more sense?  Though that still sounds like something that should go into the `.typemap`...?)